README.md
# count_down_timer

------
A new Flutter  countdown project that can be used to send SMS verification code scenarios


------

## How to Use

```python
# add this line to your dependencies
count_down_timer: ^0.0.6
```

```python
import 'package:count_down_timer/count_down_timer.dart';

```

```python
CountDownWidget(
        startSeconds: 6,
        onTapCallback: (timer) {
          print('onTapCallback：current is ${timer.tick}s');
        },
        language: CountDown.EN,
        verifyStr: "Send",
        enableTS: TextStyle(color: const Color(0xff000000)),
        disableTS: TextStyle(color: const Color(0xff999999)),
      )
```
### API

| property        | description    | defaultValue
| :--------   | :-----  |:--------
| startSeconds     | int    | 60
| onTapCallback        | Function （timer） |
| enableTS        | TextStyle ： clickable TextStyle|TextStyle(color: const Color(0xff000000)) |
| disableTS        | TextStyle ： unclickable TextStyle  |TextStyle(fontSize: 16, color: const Color(0xff999999)) |
| verifyStr        | String   |
| language        | CountDown.EN or CountDown.ZH |


###  [GitHub](https://github.com/manburenshenglu/count_down_widget)





CHANGELOG.md
* TODO: Describe initial release.
## [0.0.1] - TODO: Add release date.
## [0.0.2] - TODO: Add release date.
## [0.0.3] - TODO: Add release date.
add property  String type: verifyStr
## [0.0.4] - TODO: Add release date.
## [0.0.5] - TODO: Add release date.
add property  enmu type: language
## [0.0.6] - TODO: release date:2019.08.10.
upgrade Dart SDK

test/count_down_test.dart
import 'package:count_down_timer/count_down_timer.dart';
import 'package:flutter/material.dart';

void main() {
  CountDownWidget(
    startSeconds: 6,
    onTapCallback: (timer) {
      print('onTapCallback：current is ${timer.tick}s');
    },
    language: CountDown.EN,
    verifyStr: "Send",
    enableTS: TextStyle(color: const Color(0xff000000)),
    disableTS: TextStyle(color: const Color(0xff999999)),
  );
}


lib/count_down.dart
library count_down;

import 'dart:async';

import 'package:flutter/widgets.dart';
import 'package:flutter/material.dart';

///author xiejianlong
///description 倒计时组件
///date 2019/6/11 14:53
///modify
class CountDownWidget extends StatefulWidget {
  final int startSeconds;
  final TextStyle enableTS, disableTS;

  /// 用户点击时的回调函数。
  final Function onTapCallback;

  CountDownWidget(
      {this.startSeconds: 60,
        this.onTapCallback,
        this.enableTS,
        this.disableTS});

  @override
  _CountDownState createState() {
    // TODO: implement createState
    return _CountDownState();
  }
}

class _CountDownState extends State<CountDownWidget> {
  Timer _timer;
  int _seconds;
  bool _enable = true;

  /// 当前墨水瓶（`InkWell`）的字体样式。
  TextStyle inkWellStyle = _defaultEnableTS;

  /// 当前墨水瓶（`InkWell`）的文本。
  String _verifyStr = '获取验证码';

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    _seconds = widget.startSeconds;
    print('sec=${_seconds}');
  }
  @override
  void dispose() {
    // TODO: implement dispose
    super.dispose();
    _cancelTimer();
  }
  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return _enable
        ? InkWell(
      child: Text(
        '${_verifyStr}',
        style: inkWellStyle,
      ),
      onTap: (_seconds == widget.startSeconds)
          ? () {
        _startTimer();
      }
          : null,
    )
        : InkWell(
      child: Text(
        '${_verifyStr}',
        style: _defaultDisableTS,
      ),
    );
  }

  void _startTimer() {
    _timer = Timer.periodic(Duration(seconds: 1), (timer) {
      if (_seconds == 0) {
        _cancelTimer();
        _seconds = widget.startSeconds;
        _verifyStr = '重新发送';
        _enable = true;
        inkWellStyle = _defaultEnableTS;
        setState(() {});
        return;
      }

      _enable = false;
      _seconds--;
      _verifyStr = '${_seconds}s';
      setState(() {});
      widget.onTapCallback(timer);
    });
  }

  void _cancelTimer() {
    _timer?.cancel();
  }
}

final TextStyle _defaultEnableTS =
TextStyle(fontSize: 16, color: const Color(0xff00ff00));
final TextStyle _defaultDisableTS =
TextStyle(fontSize: 16, color: const Color(0xff999999));



lib/count_down_timer.dart
library count_down_timer;

import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

///author xiejianlong
///description count down widget
///date 2019/6/11 14:53
///modify

enum CountDown { EN, ZH }

class CountDownWidget extends StatefulWidget {
  final int startSeconds;
  final TextStyle enableTS, disableTS;
  final String verifyStr;
  final CountDown language;

  /// click callback
  final Function onTapCallback;

  CountDownWidget(
      {this.startSeconds: 60,
        this.onTapCallback,
        this.enableTS:
        const TextStyle(fontSize: 16, color: const Color(0xff00ff00)),
        this.disableTS:
        const TextStyle(fontSize: 16, color: const Color(0xff999999)),
        this.verifyStr,
        this.language: CountDown.ZH});

  @override
  _CountDownState createState() {
    // TODO: implement createState
    return _CountDownState();
  }
}

class _CountDownState extends State<CountDownWidget> {
  Timer _timer;
  int _seconds;
  bool _enable = true;

  TextStyle _inkWellStyle;

  String _verifyStr;

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    _verifyStr = widget.language == CountDown.ZH ? '获取验证码' : 'Send';
    _seconds = widget.startSeconds;
    _inkWellStyle = widget.enableTS;
  }

  @override
  void dispose() {
    // TODO: implement dispose
    super.dispose();
    _cancelTimer();
  }

  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return InkWell(
      child: Text(
        '$_verifyStr',
        style: _inkWellStyle,
      ),
      onTap: (_enable && (_seconds == widget.startSeconds))
          ? () {
        _startTimer();
      }
          : null,
    );
  }

  void _startTimer() {
    _timer = Timer.periodic(Duration(seconds: 1), (timer) {
      if (_seconds == 0) {
        _cancelTimer();
        _seconds = widget.startSeconds;
        _verifyStr = widget.language == CountDown.ZH ? '重新发送' : 'Resend';
        _enable = true;
        _inkWellStyle = widget.enableTS;
        setState(() {});
        return;
      }

      _enable = false;
      _inkWellStyle = widget.disableTS;
      _seconds--;
      _verifyStr = '${_seconds}s';
      setState(() {});
      widget.onTapCallback(timer);
    });
  }

  void _cancelTimer() {
    _timer?.cancel();
  }
}


example/example.dart
import 'package:flutter/material.dart';

import 'count_down_timer.dart';

///author xiejianlong
///description
///date 2019/6/13 17:28
///modify

void main() => runApp(Example());

class Example extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'CountDownWidget Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: CountDownWidget(
        startSeconds: 6,
        onTapCallback: (timer) {
          print('onTapCallback：current is ${timer.tick}s');
        },
        language: CountDown.EN,
        verifyStr: "Send",
        enableTS: TextStyle(color: const Color(0xff000000)),
        disableTS: TextStyle(color: const Color(0xff999999)),
      ),
    );
  }
}
